k8s.gcr.io/lib.sh

```diff
@@ -0,0 +1,279 @@
+        return 1
+    fi
+    region="$1"
+
+    if [ -z "${region}" ]; then
+        echo "gcr.io"
+    else
+        echo "${region}.gcr.io"
+    fi
+}
+
+# Get the service account email for a given short name
+# $1: The GCP project
+# $2: The name
+function svc_acct_for() {
+    if [ $# != 2 -o -z "$1" -o -z "$2" ]; then
+        echo "svc_acct_for(project, name) requires 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    name="$2"
+
+    echo "${name}@${project}.iam.gserviceaccount.com"
+}
+
+# Ensure that a project exists in our org.
+# $1: The GCP project
+function ensure_project() {
+    if [ $# != 1 -o -z "$1" ]; then
+        echo "ensure_project(project) requires 1 argument" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+
+    if ! gcloud projects describe "${project}" >/dev/null 2>&1; then
+        gcloud projects create "${project}" \
+            --organization "${GCP_ORG}"
+    else
+        o=$(gcloud projects \
+                describe "${project}" \
+                --flatten='parent[]' \
+                --format='csv[no-heading](type, id)' \
+                | grep ^organization \
+                | cut -f2 -d,)
+        if [ "$o" != "${GCP_ORG}" ]; then
+            echo "project ${project} exists, but not in our org: got ${o}" > /dev/stderr
+            return 2
+        fi
+    fi
+}
+
+# Link a project to our billing account
+# $1: The GCP project
+function ensure_billing() {
+    if [ $# != 1 -o -z "$1" ]; then
+        echo "ensure_billing(project) requires 1 argument" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+
+    gcloud beta billing projects link "${project}" \
+        --billing-account "${GCP_BILLING}"
+}
+
+# Enable an API
+# $1: The GCP project
+# $2: The API (e.g. containerregistry.googleapis.com)
+function enable_api() {
+    if [ $# != 2 -o -z "$1" -o -z "$2" ]; then
+        echo "enable_api(project, api) requires 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    api="$2"
+
+    gcloud --project "${project}" services enable "${api}"
+}
+
+# Ensure the bucket backing the repo exists and is world-readable
+# $1: The GCP project
+function ensure_repo() {
+    if [ $# -lt 1 -o $# -gt 2 -o -z "$1" ]; then
+        echo "ensure_repo(project, [region]) requires 1 or 2 arguments" > /dev/stderr
+        return 1
+    fi
+    project="$1"
+    region="${2:-}"
+
+    if ! gsutil ls $(gcs_bucket_for "${project}" "${region}") >/dev/null 2>&1; then
+        host=$(gcr_host_for "${region}")
+        image="ceci-nest-pas-une-image"
+        dest="${host}/${project}/${image}"
+        docker pull k8s.gcr.io/pause
+        docker tag k8s.gcr.io/pause "${dest}"
+        docker push "${dest}"
+        gcloud --project "${project}" \
+            container images delete --quiet "${dest}:latest"
+    fi
+
+    gsutil iam ch allUsers:objectViewer $(gcs_bucket_for "${project}" "${region}")
```

Comment:
Since $region can be an empty string (line 138), we should only expand to a double-quoted var if it is not empty. So I think

```
${region:+"$region"}
```

would be better. Then from `gcs_bucket_for`s perspective, the region, if there, is guaranteed to be non-empty, which would simplify the validation logic on line 33.

Also I would move the call to `gcs_bucket_for` here to its own line, assign to a variable and use that instead of doing it inline like this.

Reply:
I explicitly check for "" in bucket_for, to keep this simple.  Alternatively, I can make the 2nd arg required, even if it is "", which doesn't seem much better.  Like, I had to go look that syntax up and test it to convince myself it works.  I'll do it if you think it's really better but I assume that is a bash-ism?

Reply:
It is actually a POSIX thing, as per https://unix.stackexchange.com/a/415992.

I admit that these POSIX variable expansions are not intuitive --- I'm fine with leaving this as-is for now.

Reply:
Interesting.  Bash docs say "If parameter is null or unset, nothing is substituted, otherwise the expansion of word is substituted.".  Testing shows `""` counts as null or unset.